File: src/test/java/app/tweditor/ThemeScreenshotTest.java
package app.tweditor;

import java.awt.Robot;
import java.awt.image.BufferedImage;
import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Properties;
import javax.imageio.ImageIO;
import javax.swing.SwingUtilities;
import javax.swing.UIManager;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.Assumptions;

import static org.junit.jupiter.api.Assertions.assertNotNull;

/**
 * Captures light/dark screenshots of the main window for theme tickets.
 * Gated behind -Dtweditor.screenshots=true so the suite never opens windows by default:
 *   gradlew test --tests "app.tweditor.ThemeScreenshotTest" -Dtweditor.screenshots=true
 */
class ThemeScreenshotTest {

  @Test
  void captureLightAndDarkMainWindow() throws Exception {
    Assumptions.assumeTrue(Boolean.getBoolean("tweditor.screenshots"));

    Path outDir = Path.of("docs", "screenshots");
    Files.createDirectories(outDir);
    capture(ThemeSelection.Preference.LIGHT, outDir.resolve("theme-light.png"));
    capture(ThemeSelection.Preference.DARK, outDir.resolve("theme-dark.png"));
  }

  private static void capture(ThemeSelection.Preference preference, Path target) throws Exception {
    initStatics();

    SwingUtilities.invokeAndWait(() -> {
      try {
        UIManager.setLookAndFeel(ThemeSelection.lookAndFeel(preference));
      } catch (Exception exc) {
        throw new RuntimeException(exc);
      }
    });

    MainWindow[] holder = new MainWindow[1];
    SwingUtilities.invokeAndWait(() -> holder[0] = new MainWindow());
    MainWindow window = holder[0];
    SwingUtilities.invokeAndWait(() -> {
      window.setAlwaysOnTop(true);
      window.setLocationRelativeTo(null);
      window.pack();
      window.setVisible(true);
      window.toFront();
    });

    loadFixture();

    SwingUtilities.invokeAndWait(() -> {
      try {
        Main.dataChanging = true;
        DBList list = (DBList) Main.database.getTopLevelStruct().getValue();
        list = (DBList) list.getElement("Mod_PlayerList").getValue();
        list = (DBList) list.getElement(0).getValue();

        window.statsPanel.setFields(list);
        window.attributesPanel.setFields(list);
        window.signsPanel.setFields(list);
        window.stylesPanel.setFields(list);
        window.questsPanel.setFields(list);
        window.difficultyPanel.setFields(list);

        window.tabbedPane.setSelectedIndex(0);
        window.tabbedPane.setVisible(true);
        Main.dataChanging = false;
        Main.dataModified = false;
      } catch (Exception exc) {
        throw new RuntimeException(exc);
      }
    });

    Thread.sleep(1000);

    Robot robot = new Robot(window.getGraphicsConfiguration().getDevice());
    BufferedImage shot = robot.createScreenCapture(window.getBounds());
    ImageIO.write(shot, "png", target.toFile());

    SwingUtilities.invokeAndWait(window::dispose);
  }

  private static void initStatics() throws Exception {
    Main.fileSeparator = System.getProperty("file.separator");
    Main.lineSeparator = System.getProperty("line.separator");
    Main.tmpDir = System.getProperty("java.io.tmpdir");
    Main.smmFile = new File(Main.tmpDir, "TWEditor.smm");
    Main.databaseFile = new File(Main.tmpDir, "TWEditor.ifo");
    Main.modFile = new File(Main.tmpDir, "TWEditor.mod");
    Main.playerFile = new File(Main.tmpDir, "TWEditor.player");
    Main.properties = new Properties();
    Main.languageID = 3;
    Main.resourceFiles = new HashMap<>();
    Main.itemTemplates = new ArrayList<>();
    Main.stringsDatabase = new StringsDatabase(fakeTlk().getPath());
    Main.dataModified = false;
    Main.dataChanging = false;
  }

  private static void loadFixture() throws Exception {
    File saveFile = SaveSeamSupport.copyFixtureTo(Files.createTempDirectory("theme-shots"));
    LoadFile task = new LoadFile(new ProgressDialog(Main.mainWindow, "Loading"), saveFile);
    task.run();
    assertNotNull(Main.saveDatabase, "fixture save failed to load");
  }

  /**
   * A minimal but valid TLK so localized-string lookups resolve to an empty
   * string instead of hitting a missing strings database.
   */
  private static File fakeTlk() {
    try {
      File file = Files.createTempFile("theme-shots", ".tlk").toFile();
      file.deleteOnExit();
      try (FileOutputStream out = new FileOutputStream(file)) {
        out.write(new byte[]{'T', 'L', 'K', ' ', 'V', '3', '.', '0'});
        out.write(leInt(3));  // language id
        out.write(leInt(1));  // string count
        out.write(leInt(20 + 40));  // string data offset
        byte[] entry = new byte[40];
        entry[0] = 0x01;  // present
        entry[28] = 0x00;  // data offset
        entry[32] = 0x00;  // data length
        out.write(entry);
        out.write(new byte[0]);
      }
      return file;
    } catch (IOException exc) {
      throw new RuntimeException(exc);
    }
  }

  private static byte[] leInt(int value) {
    return new byte[]{
        (byte) value, (byte) (value >>> 8), (byte) (value >>> 16), (byte) (value >>> 24)
    };
  }
}

File: src/test/kotlin/app/tweditor/ThemeScreenshotTest.kt
package app.tweditor

import java.awt.Robot
import java.awt.image.BufferedImage
import java.io.File
import java.io.FileOutputStream
import java.nio.file.Files
import java.nio.file.Path
import javax.imageio.ImageIO
import javax.swing.SwingUtilities
import org.junit.jupiter.api.Assumptions.assumeTrue
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.Assertions.assertNotNull

class ThemeScreenshotTest {

    @Test
    fun captureLightAndDarkMainWindow() {
        assumeTrue(java.lang.Boolean.getBoolean("tweditor.screenshots"))

        val outDir = Path.of("docs", "screenshots")
        Files.createDirectories(outDir)
        capture(ThemeSelection.Preference.LIGHT, outDir.resolve("theme-light.png"))
        capture(ThemeSelection.Preference.DARK, outDir.resolve("theme-dark.png"))
    }

    private fun capture(preference: ThemeSelection.Preference, target: Path) {
        val environment = createEnvironment()

        SwingUtilities.invokeAndWait {
            ThemeSelection.install(preference)
        }

        var window: MainWindow? = null
        SwingUtilities.invokeAndWait {
            window = MainWindow(environment)
        }
        SwingUtilities.invokeAndWait {
            window!!.isAlwaysOnTop = true
            window!!.setLocationRelativeTo(null)
            window!!.pack()
            window!!.isVisible = true
            window!!.toFront()
        }

        loadFixture(environment, window!!)

        SwingUtilities.invokeAndWait {
            try {
                window!!.session.setDataChanging(true)
                var list = window!!.session.database!!.getTopLevelStruct()!!.getValue() as DBList
                list = list.getElement("Mod_PlayerList")!!.getValue() as DBList
                list = list.getElement(0).getValue() as DBList

                window!!.statsPanel.setFields(list)
                window!!.attributesPanel.setFields(list)
                window!!.signsPanel.setFields(list)
                window!!.stylesPanel.setFields(list)
                window!!.questsPanel.setFields(list)
                window!!.difficultyPanel.setFields(list)

                window!!.tabbedPane.selectedIndex = 0
                window!!.tabbedPane.isVisible = true
                window!!.session.setDataChanging(false)
                window!!.session.setDataModified(false)
            } catch (exc: Exception) {
                throw RuntimeException(exc)
            }
        }

        Thread.sleep(1000)

        val robot = Robot(window!!.graphicsConfiguration.device)
        val shot = robot.createScreenCapture(window!!.bounds)
        ImageIO.write(shot, "png", target.toFile())

        SwingUtilities.invokeAndWait {
            window!!.dispose()
        }
    }

    private fun createEnvironment(): AppEnvironment {
        val environment = AppEnvironment()
        environment.fileSeparator = System.getProperty("file.separator")
        environment.lineSeparator = System.getProperty("line.separator")
        environment.tmpDir = System.getProperty("java.io.tmpdir")
        environment.properties = java.util.Properties()
        environment.languageID = 3
        environment.resourceFiles = HashMap()
        environment.itemTemplates = ArrayList()
        environment.stringsDatabase = StringsDatabase(fakeTlk().getPath())
        return environment
    }

    private fun loadFixture(environment: AppEnvironment, window: MainWindow) {
        val saveFile = SaveSeamSupport.copyFixtureTo(Files.createTempDirectory("theme-shots"))
        val task = LoadFile(ProgressDialog(window, "Loading"), window.session, environment, saveFile)
        task.run()
        assertNotNull(window.session.saveDatabase, "fixture save failed to load")
    }

    /**
     * A minimal but valid TLK so localized-string lookups resolve to an empty
     * string instead of hitting a missing strings database.
     */
    private fun fakeTlk(): File {
        val file = Files.createTempFile("theme-shots", ".tlk").toFile()
        file.deleteOnExit()
        FileOutputStream(file).use { out ->
            out.write(byteArrayOf('T'.code.toByte(), 'L'.code.toByte(), 'K'.code.toByte(), ' '.code.toByte(), 'V'.code.toByte(), '3'.code.toByte(), '.'.code.toByte(), '0'.code.toByte()))
            out.write(leInt(3))
            out.write(leInt(1))
            out.write(leInt(20 + 40))
            val entry = ByteArray(40)
            entry[0] = 0x01
            entry[28] = 0x00
            entry[32] = 0x00
            out.write(entry)
        }
        return file
    }

    private fun leInt(value: Int): ByteArray {
        return byteArrayOf(
            value.toByte(),
            (value ushr 8).toByte(),
            (value ushr 16).toByte(),
            (value ushr 24).toByte()
        )
    }
}
